### ISSUE TITLE
method call on nil object and suspicious  connection closing behaviour

### ISSUE THREAD

**Issue by stefantalpalaru** [status-im/nim-libp2p#89]
Intermittent exception, intermittently caught or not, while running `make LOG_LEVEL=TRACE testnet1` on nim-beacon-chain's devel branch:

```text
TRC 2020-02-23 13:25:13+01:00 closing connection                         tid=25059
TRC 2020-02-23 13:25:13+01:00 sending encrypted bytes                    tid=25059 bytes=0C00 topic=secio
TRC 2020-02-23 13:25:13+01:00 Writing message                            tid=25059 message=00000022BF1372D352F967D029DCE74FE2B7950A38383425275552433F431C6D0670DEEE93EE topic=secio
TRC 2020-02-23 13:25:13+01:00 Recieved message header                    tid=25059 header=00000022 length=34 topic=secio
TRC 2020-02-23 13:25:13+01:00 Received message body                      tid=25059 buffer=BF1372D352F967D029DCE74FE2B7950A38383425275552433F431C6D0670DEEE93EE length=34 topic=secio
TRC 2020-02-23 13:25:13+01:00 waiting for data                           tid=25059 topic=Mplex
TRC 2020-02-23 13:25:13+01:00 connection closed                          tid=25059 closed=false
TRC 2020-02-23 13:25:13+01:00 read public key from message               tid=25059 pubKey="Secp256k1 key (0381C11F59A277AF11F80924DCEF507A1E563706899B476AEEF34C2831066E3CC5)" topic=identify
TRC 2020-02-23 13:25:13+01:00 read address bytes from message            tid=25059 address=/ip4/0.0.0.0/tcp/9000 topic=identify
TRC 2020-02-23 13:25:13+01:00 read protoVersion from message             tid=25059 protoVersion=ipfs/0.1.0 topic=identify
TRC 2020-02-23 13:25:13+01:00 read agentVersion from message             tid=25059 agentVersion=nim-libp2p/0.0.1 topic=identify
TRC 2020-02-23 13:25:13+01:00 Identify for remote peer succeded          tid=25059 topic=identify
TRC 2020-02-23 13:25:13+01:00 connection's peerInfo                      tid=25059 peerInfo="PeerID: 16Uiu2HAmMPVjzko7GiQjg5bd5SsagcAzCb8GMSTdKUnW6vM8eTfA\nPeer Addrs: /ip4/0.0.0.0/tcp/9000\n" topic=Switch
TRC 2020-02-23 13:25:13+01:00 adding muxer for peer                      tid=25059 peer=16Uiu2HAmMPVjzko7GiQjg5bd5SsagcAzCb8GMSTdKUnW6vM8eTfA topic=Switch
TRC 2020-02-23 13:25:13+01:00 identify: identified remote peer           tid=25059 peer=16Uiu2HAmMPVjzko7GiQjg5bd5SsagcAzCb8GMSTdKUnW6vM8eTfA topic=Switch
TRC 2020-02-23 13:25:13+01:00 read header varint                         tid=25059 topic=MplexCoder varint=12
TRC 2020-02-23 13:25:13+01:00 read data len varint                       tid=25059 topic=MplexCoder varint=0
TRC 2020-02-23 13:25:13+01:00 closing connection                         tid=25059
TRC 2020-02-23 13:25:13+01:00 sending encrypted bytes                    tid=25059 bytes=0C00 topic=secio
TRC 2020-02-23 13:25:13+01:00 Writing message                            tid=25059 message=00000022FC81E99EA84B5504B122454DE9B1503944F43A4F6E34A3BE0DEB3239BBC7D2B8662C topic=secio
TRC 2020-02-23 13:25:13+01:00 Dialing peer                               tid=25059 peer=16Uiu2HAmGF43WzRjXm9kLKHdDCYYQsjViwyvrPY9WpZ5tgnSkbbF topic=Switch
TRC 2020-02-23 13:25:13+01:00 Reusing existing connection                tid=25059 topic=Switch
TRC 2020-02-23 13:25:13+01:00 initiating handshake                       tid=25059 codec="\x13/multistream/1.0.0\n" topic=Multistream
TRC 2020-02-23 13:25:13+01:00 read message from connection               tid=25059 data=@[] id=1 msgType=CloseOut topic=Mplex
TRC 2020-02-23 13:25:13+01:00 picking remote channels                    tid=25059 initiator=false topic=Mplex
TRC 2020-02-23 13:25:13+01:00 closing channel                            tid=25059 id=1 initiator=false msgType=CloseOut topic=Mplex
TRC 2020-02-23 13:25:13+01:00 picking remote channels                    tid=25059 initiator=false topic=Mplex
TRC 2020-02-23 13:25:13+01:00 waiting for data                           tid=25059 topic=Mplex
FAT 2020-02-23 13:25:13+01:00 Fatal exception reached                    tid=25059 err="cannot dispatch; dispatcher is nil\nAsync traceback:\nException message: cannot dispatch; dispatcher is nil\nException type:"
```

Notice that the stream appears to not be closed here: `connection closed                          tid=25059 closed=false`. Giovanni thinks this might stem from a Nim runtime bug where methods are dispatched to the parent class (which would be LPStream, with a close() method that would fail with an assertion error).

**Comment by sinkingsugar** [comment]
Had the same issue with `Connection` indeed, I worked around it tho in the end.
https://github.com/nim-lang/Nim/issues/12673 likely.

**Comment by dryajov** [comment]
resolved in #93
